src/Caching/Impl/ModuleDatabase.cs

```diff
@@ -17,151 +17,117 @@
 using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.IO;
-using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using LiteDB;
 using Microsoft.Python.Analysis.Analyzer;
+using Microsoft.Python.Analysis.Caching.IO;
 using Microsoft.Python.Analysis.Caching.Models;
-using Microsoft.Python.Analysis.Dependencies;
 using Microsoft.Python.Analysis.Modules;
 using Microsoft.Python.Analysis.Types;
 using Microsoft.Python.Core;
 using Microsoft.Python.Core.IO;
 using Microsoft.Python.Core.Logging;
-using Microsoft.Python.Parsing.Ast;
+using Microsoft.Python.Core.Services;
 
 namespace Microsoft.Python.Analysis.Caching {
-    internal sealed class ModuleDatabase : IModuleDatabaseService {
-        private readonly ConcurrentDictionary<string, IDependencyProvider> _dependencies = new ConcurrentDictionary<string, IDependencyProvider>();
+    internal sealed class ModuleDatabase : IModuleDatabaseService, IDisposable {
+        private readonly object _modulesLock = new object();
+        private readonly Dictionary<string, PythonDbModule> _modulesCache
+            = new Dictionary<string, PythonDbModule>();
+
+        private readonly ConcurrentDictionary<string, ModuleModel> _modelsCache
+            = new ConcurrentDictionary<string, ModuleModel>();
+
+        private readonly ConcurrentDictionary<AnalysisModuleKey, bool> _searchResults
+            = new ConcurrentDictionary<AnalysisModuleKey, bool>();
 
         private readonly IServiceContainer _services;
         private readonly ILogger _log;
         private readonly IFileSystem _fs;
-
-        public ModuleDatabase(IServiceContainer services) {
-            _services = services;
-            _log = services.GetService<ILogger>();
-            _fs = services.GetService<IFileSystem>();
-
-            var cfs = services.GetService<ICacheFolderService>();
-            CacheFolder = Path.Combine(cfs.CacheFolder, $"{CacheFolderBaseName}{DatabaseFormatVersion}");
+        private readonly AnalysisCachingLevel _defaultCachingLevel;
+        private readonly CacheWriter _cacheWriter;
+        private AnalysisCachingLevel? _cachingLevel;
+
+        public ModuleDatabase(IServiceManager sm, string cacheFolder = null) {
+            _services = sm;
+            _log = _services.GetService<ILogger>();
+            _fs = _services.GetService<IFileSystem>();
+            _defaultCachingLevel = AnalysisCachingLevel.Library;
```

Comment:
What makes this level differ from the other default?

Reply:
It is mostly for tests so default is full functionality. If not provided, like in tests, `Library` is used since there are now tests for full restore, i.e. analysis - write cache, reset analyzer - reload. If provided, such as from settings in the real case, the provided one will be used.